File: py-langgraph/src/adapters/registry.py
"""
指令注册中心
"""

from typing import Callable, Awaitable, Dict, List, Optional
from .qq import NormalizedMessage


class CommandRegistry:
    """指令注册中心 - 单例"""

    _instance = None
    _commands: Dict[str, Dict] = {}

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def register(self, name: str, description: str):
        """装饰器：注册指令"""
        def decorator(func: Callable[[NormalizedMessage, List[str]], Awaitable[str]]):
            self._commands[name] = {
                "name": name,
                "description": description,
                "handler": func,
            }
            return func
        return decorator

    def register_handler(self, name: str, description: str, handler):
        """直接注册处理器"""
        self._commands[name] = {
            "name": name,
            "description": description,
            "handler": handler,
        }

    async def dispatch(self, msg: NormalizedMessage) -> Optional[str]:
        """分发指令"""
        parts = msg.content.strip().split()
        if not parts:
            return None

        command = parts[0].lower()
        args = parts[1:]

        entry = self._commands.get(command)
        if entry:
            return await entry["handler"](msg, args)

        return None

    def list_commands(self) -> List[Dict[str, str]]:
        """列出所有指令"""
        return [
            {"name": c["name"], "description": c["description"]}
            for c in self._commands.values()
        ]


# 全局实例
registry = CommandRegistry()

File: ts-langchain/src/adapters/commands/registry.ts
/**
 * 指令注册中心
 * 装饰器模式注册指令
 */

import { NormalizedMessage } from "../types.js";

export type CommandHandler = (msg: NormalizedMessage, args: string[]) => Promise<string>;

interface CommandEntry {
  name: string;
  description: string;
  handler: CommandHandler;
}

class CommandRegistryClass {
  private commands: Map<string, CommandEntry> = new Map();

  register(name: string, description: string) {
    return function (target: any, propertyKey: string, descriptor: PropertyDescriptor) {
      const originalMethod = descriptor.value;
      descriptor.value = async function (...args: any[]) {
        return originalMethod.apply(target, args);
      };
      registry.registerHandler(name, description, descriptor.value);
      return descriptor;
    };
  }

  registerHandler(name: string, description: string, handler: CommandHandler) {
    this.commands.set(name, { name, description, handler });
  }

  async dispatch(msg: NormalizedMessage): Promise<string | null> {
    const parts = msg.content.trim().split(/\s+/);
    const command = parts[0].toLowerCase();
    const args = parts.slice(1);

    const entry = this.commands.get(command);
    if (entry) {
      return entry.handler(msg, args);
    }

    return null; // 不是指令，交给 Agent 处理
  }

  list(): Array<{ name: string; description: string }> {
    return Array.from(this.commands.values()).map((c) => ({
      name: c.name,
      description: c.description,
    }));
  }
}

export const registry = new CommandRegistryClass();
